### PR TITLE
Dispute modifications

### PR DESCRIPTION
**Pull request by Spablob**
Dispute adjustments
- [x] add onResolveDispute hook for future-proofing
- [x] add new mechanic of permissionless self-tagging
- [x] add multicall inheritance (to facilitate batch call for tagging)

Other fixes
- [x] add missing zero variable initializer/constructor restrictions

### CODE REVIEW

File: contracts/modules/dispute/DisputeModule.sol [hunk 1]
```diff
@@ -239,19 +260,69 @@ contract DisputeModule is
         emit DisputeCancelled(disputeId, data);
     }
 
+    /// @notice Tags a derivative if a parent has been tagged with an infringement tag
+    /// @param parentIpId The infringing parent ipId
+    /// @param derivativeIpId The derivative ipId
+    /// @param parentDisputeId The dispute id that tagged the parent ipId as infringing
+    function tagDerivativeIfParentInfringed(
```
Comment (kingster-will) [inline contracts/modules/dispute/DisputeModule.sol:267]: Who is supposed to execute this function?

Comment (Spablob) [reply contracts/modules/dispute/DisputeModule.sol:267]: it's permissionless

**Review by LeoHChen — CHANGES_REQUESTED** [review]
This one should be split into two PRs.  One for Dispute modification and one for Royalty fix.

File: Makefile [hunk 2]
```diff
@@ -44,8 +44,8 @@ format:
 coverage:
 	mkdir -p coverage
 	forge coverage --report lcov
-	lcov --remove lcov.info -o coverage/lcov.info 'test/*' 'script/*' --rc branch_coverage=1
```
Comment (Ramarti) [inline Makefile:47]: Should this change be here?

File: contracts/modules/dispute/DisputeModule.sol [hunk 3]
```diff
@@ -239,19 +260,69 @@ contract DisputeModule is
...
     /// @notice Resolves a dispute after it has been judged
     /// @param disputeId The dispute id
-    function resolveDispute(uint256 disputeId) external {
+    /// @param data The data to resolve the dispute
+    function resolveDispute(uint256 disputeId, bytes calldata data) external {
         DisputeModuleStorage storage $ = _getDisputeModuleStorage();
         Dispute memory dispute = $.disputes[disputeId];
 
-        if (msg.sender != dispute.disputeInitiator) revert Errors.DisputeModule__NotDisputeInitiator();
+        if (dispute.parentDisputeId == 0 && msg.sender != dispute.disputeInitiator)
```
Comment (jdubpark) [inline contracts/modules/dispute/DisputeModule.sol:313]: Why check `dispute.parentDisputeId == 0`? 

If it's 1 or more, the branch skips on checking if msg.sender is the dispute initiator.

File: contracts/lib/DataUniqueness.sol
Comment (jdubpark) [inline contracts/lib/DataUniqueness.sol:1]: Unrelated changes, I'd recommend a separate PR next time.

Comment (LeoHChen) [reply contracts/lib/DataUniqueness.sol:1]: good point, actually there was a PR.  shouldn't merge those PR together.

Comment (LeoHChen) [reply contracts/lib/DataUniqueness.sol:1]: reopened PR #65

File: contracts/modules/dispute/DisputeModule.sol [hunk 4]
```diff
@@ -239,19 +260,69 @@ contract DisputeModule is
...
+        address arbitrationPolicy = $.arbitrationPolicies[derivativeIpId];
+        if (!$.isWhitelistedArbitrationPolicy[arbitrationPolicy]) arbitrationPolicy = $.baseArbitrationPolicy;
+
+        uint256 disputeId = ++$.disputeCounter;
+
+        $.disputes[disputeId] = Dispute({
+            targetIpId: derivativeIpId,
+            disputeInitiator: msg.sender,
+            arbitrationPolicy: arbitrationPolicy,
+            linkToDisputeEvidence: "",
```
Comment (LeoHChen) [inline contracts/modules/dispute/DisputeModule.sol:290]: no evidence needed? or you want to leave a note here saying it's due to parent infringement?

File: contracts/interfaces/modules/dispute/IDisputeModule.sol [hunk 5]
```diff
@@ -181,9 +197,20 @@ interface IDisputeModule {
...
+        address derivativeIpId,
+        uint256 parentDisputeId
+    ) external;
+
     /// @notice Resolves a dispute after it has been judged
-    /// @param disputeId The dispute id
-    function resolveDispute(uint256 disputeId) external;
+    /// @param disputeId The dispute
+    /// @param data The data to resolve the dispute
+    function resolveDispute(uint256 disputeId, bytes calldata data) external;
```
Comment (LeoHChen) [inline contracts/interfaces/modules/dispute/IDisputeModule.sol:213]: Is there a recommendation of the data format here?

File: contracts/modules/royalty/policies/RoyaltyPolicyLAP.sol [hunk 6]
```diff
@@ -72,9 +72,11 @@ contract RoyaltyPolicyLAP is IRoyaltyPolicyLAP, AccessManagedUpgradeable, Reentr
     }
 
     /// @notice Initializer for this implementation contract
-    /// @param governance The governance address
-    function initialize(address governance) external initializer {
-        __AccessManaged_init(governance);
+    /// @param accessManager The address of the protocol admin roles contract
+    function initialize(address accessManager) external initializer {
+        if (accessManager == address(0)) revert Errors.RoyaltyPolicyLAP__ZeroAccessManager();
```
Comment (LeoHChen) [inline contracts/modules/royalty/policies/RoyaltyPolicyLAP.sol:77]: Do we need to do this check for all other modules?  @Ramarti ?

File: Makefile (hunk 2 above)
Comment (Spablob) [reply Makefile:47]: You mean coverage in general? cc @jdubpark 
The change was made to ensure that the branch coverage would also show up for more users - as the previous cmds were not working for me

File: contracts/modules/dispute/DisputeModule.sol (hunk 3 above)
Comment (Spablob) [reply contracts/modules/dispute/DisputeModule.sol:313]: Yes, so there are 2 types of dispute now:

1. the normal one that requires judgment: this one only the dispute initiator can resolve
2. the automatic one: this one is permissionless to tag and permissionless to resolve

the way to distinguish between them is whether the `dispute.parentDisputeId == 0`

File: contracts/modules/dispute/DisputeModule.sol (hunk 4 above)
Comment (Spablob) [reply contracts/modules/dispute/DisputeModule.sol:290]: No evidence is needed to tag a derivative if the parent was already tagged. 

There is no `linkToDisputeEvidence` to save gas. The evidence that justified the tagging of the parent is accessible via the `parentDisputeId`.

File: contracts/interfaces/modules/dispute/IDisputeModule.sol (hunk 5 above)
Comment (Spablob) [reply contracts/interfaces/modules/dispute/IDisputeModule.sol:213]: For the current arbitration policy SP - we can pass we don't need to pass anything - just "".

The was added for future-proofing.

File: contracts/modules/royalty/policies/RoyaltyPolicyLAP.sol (hunk 6 above)
Comment (Spablob) [reply contracts/modules/royalty/policies/RoyaltyPolicyLAP.sol:77]: Would say that if so - perhaps we can add it in a separate pr @Ramarti

**Comment by Spablob** [comment]
> This one should be split into two PRs. One for Dispute modification and one for Royalty fix.

Removed royalty-related fixes from the pr in https://github.com/storyprotocol/protocol-core-v1/pull/60/commits/1520309a4cbcb848200c13ce3f63922314e502a5

File: contracts/modules/dispute/DisputeModule.sol (hunk 3 above)
Comment (LeoHChen) [reply contracts/modules/dispute/DisputeModule.sol:313]: you need to add better comments here to avoid confusion later, especially this is a key business logic. thanks.

File: contracts/interfaces/modules/dispute/IDisputeModule.sol [hunk 7]
```diff
@@ -181,9 +197,20 @@ interface IDisputeModule {
...
+        address parentIpId,
+        address derivativeIpId,
+        uint256 parentDisputeId
+    ) external;
+
     /// @notice Resolves a dispute after it has been judged
-    /// @param disputeId The dispute id
-    function resolveDispute(uint256 disputeId) external;
+    /// @param disputeId The dispute
+    /// @param data The data to resolve the dispute
```
Comment (LeoHChen) [inline contracts/interfaces/modules/dispute/IDisputeModule.sol:212]: please add a todo here, the data format is not defined yet, maybe add another github issue for discussion

File: contracts/modules/dispute/DisputeModule.sol [hunk 8]
```diff
@@ -239,19 +260,69 @@ contract DisputeModule is
...
+    function tagDerivativeIfParentInfringed(
+        address parentIpId,
+        address derivativeIpId,
+        uint256 parentDisputeId
+    ) external {
+        DisputeModuleStorage storage $ = _getDisputeModuleStorage();
+
+        Dispute memory parentDispute = $.disputes[parentDisputeId];
+        if (parentDispute.targetIpId != parentIpId) revert Errors.DisputeModule__ParentIpIdMismatch();
+        if (parentDispute.currentTag == IN_DISPUTE || parentDispute.currentTag == bytes32(0))
```
Comment (LeoHChen) [inline contracts/modules/dispute/DisputeModule.sol:276]: you probably need to have an explicit check on the tag?

File: contracts/interfaces/modules/dispute/IDisputeModule.sol (hunk 7 above)
Comment (Spablob) [reply contracts/interfaces/modules/dispute/IDisputeModule.sol:212]: The data format will vary with dispute policy. Each data format will be defined in each dispute policy contract and not in DisputeModule.sol contract

File: contracts/modules/dispute/DisputeModule.sol (hunk 8 above)
Comment (Spablob) [reply contracts/modules/dispute/DisputeModule.sol:276]: Wasn't clear - what did you mean?

Comment (LeoHChen) [reply contracts/modules/dispute/DisputeModule.sol:276]: you checked the currentTag == IN_DISPUTE, what's the value of the currentTag if the dispute is succeeded?  Do you need to check this condition in this function here?

File: contracts/modules/dispute/DisputeModule.sol (hunk 3 above)
Comment (Spablob) [reply contracts/modules/dispute/DisputeModule.sol:316]: addressed in https://github.com/storyprotocol/protocol-core-v1/pull/60/commits/6581549f82c44095d31241a4ac7c12f00dc13ddb

File: contracts/modules/dispute/DisputeModule.sol (hunk 8 above)
Comment (Spablob) [reply contracts/modules/dispute/DisputeModule.sol:276]: 1. while in dispute - currentTag == IN_DISPUTE
2. if a dispute succeeds then the tag can be anything SP whitelisted - ie. "PLAGIARISM".
3. if a dispute fails then tag = address(0)

Derivatives can only be tagged in a permissionless way if the parent already has a tag like "PLAGIARISM"

Comment (LeoHChen) [reply contracts/modules/dispute/DisputeModule.sol:276]: what's the default value of currentTag?

Comment (LeoHChen) [reply contracts/modules/dispute/DisputeModule.sol:276]: do you want to add a helper function to check if the dispute is not succeeded yet?

**Review by LeoHChen — APPROVED** [review]

**Review by kingster-will — APPROVED** [review]

File: contracts/modules/dispute/DisputeModule.sol (hunk 8 above)
Comment (Spablob) [reply contracts/modules/dispute/DisputeModule.sol:276]: was double-checking just now - given this check is only used one - it would be strange to create an internal function just for it. Will add comments to make it more clear.

Comment (Spablob) [reply contracts/modules/dispute/DisputeModule.sol:279]: https://github.com/storyprotocol/protocol-core-v1/pull/60/commits/3f73ef2cf048912644bbe8252e2d285900d3b2fd

**Review by LeoHChen — APPROVED** [review]
